system-addon/content-src/components/ComponentPerfTimer/ComponentPerfTimer.jsx

```diff
@@ -64,40 +66,66 @@ class TopSitesPerfTimer extends React.PureComponent {
   }
 
   _maybeSendPaintedEvent() {
-    // We don't want this to ever happen, but sometimes it does.  And when it
-    // does (typically on the first newtab at startup time calling
-    // componentDidMount), the paint(s) we care about will be later (eg
-    // in a subsequent componentDidUpdate).
-    if (!this.props.TopSites.initialized) {
-      // XXX should send bad event
+    if (!RECORDED_SECTIONS.includes(this.props.id)) {
       return;
     }
 
-    // If we've already handled a timestamp, don't do it again
+    if (!this.props.initialized) {
+      // Remember to report back when data is available.
+      this._reportMissingData = true;
+    } else if (this._reportMissingData) {
+      this._reportMissingData = false;
+      // Report that data is available later than first render call.
+      this._afterFramePaint(this._sendBadStateEvent);
+    }
+
+    // Only record first call to render.
     if (this._timestampHandled) {
       return;
     }
 
-    // And if we haven't, we're doing so now, so remember that. Even if
-    // something goes wrong in the callback, we can't try again, as we'd be
-    // sending back the wrong data, and we have to do it here, so that other
-    // calls to this method while waiting for the next frame won't also try to
-    // handle handle it.
     this._timestampHandled = true;
-
     this._afterFramePaint(this._sendPaintedEvent);
   }
 
+  _sendBadStateEvent() {
+    const dataReadyKey = `${this.props.id}_data_ready_ts`;
```

Comment:
nit: for each of these dynamically generated keys, could you add a comment above with the expected names:
`// topsites_data_ready_ts, highlights_data_ready_ts`

I know it seems a bit silly, but it's so that it's easier to find/`grep` for where these id/keys are made. Where someone not familiar with the code wouldn't know to search for `_data_ready_ts`.